clients/venice-thin-client/src/main/java/com/linkedin/venice/client/schema/RouterBackedSchemaReader.java

```diff
@@ -300,6 +290,294 @@ private SchemaEntry fetchSingleSchema(String requestPath, boolean isValueSchema)
+        DerivedSchemaEntry derivedSchemaEntry = fetchUpdateSchemaEntryFromRouter(id);
+        if (derivedSchemaEntry == null) {
+          return NOT_EXIST_UPDATE_SCHEMA_ENTRY;
+        }
+        return derivedSchemaEntry;
+      });
+    }
+  }
+
+  private SchemaEntry fetchValueSchemaEntryFromRouter(int valueSchemaId) {
+    String requestPath = TYPE_VALUE_SCHEMA + "/" + storeName + "/" + valueSchemaId;
+    return fetchSingleSchema(requestPath, true);
+  }
+
+  private DerivedSchemaEntry fetchUpdateSchemaEntryFromRouter(int valueSchemaId) {
+    String requestPath = TYPE_UPDATE_SCHEMA + "/" + storeName + "/" + valueSchemaId;
+    SchemaResponse schemaResponse;
+    try {
+      schemaResponse = fetchSingleSchemaResponse(requestPath);
+    } catch (Exception e) {
+      return null;
+    }
+    if (schemaResponse == null) {
+      return null;
+    }
+    return new DerivedSchemaEntry(
+        schemaResponse.getId(),
+        schemaResponse.getDerivedSchemaId(),
+        schemaResponse.getSchemaStr());
+  }
+
+  private Set<Integer> fetchAllValueSchemaIdsFromRouter() {
+    String requestPath = TYPE_VALUE_SCHEMA_ID + "/" + storeName;
+    MultiSchemaIdResponse multiSchemaIdResponse = fetchMultiSchemaIdResponse(requestPath);
+    if (multiSchemaIdResponse == null) {
+      return Collections.emptySet();
+    }
+    if (multiSchemaIdResponse.getSuperSetSchemaId() != SchemaData.INVALID_VALUE_SCHEMA_ID) {
+      supersetSchemaIdAtomic.set(multiSchemaIdResponse.getSuperSetSchemaId());
+    }
+    return multiSchemaIdResponse.getSchemaIdSet();
+  }
+
+  private List<SchemaEntry> fetchAllValueSchemaEntriesFromRouter() {
+    String requestPath = TYPE_VALUE_SCHEMA + "/" + storeName;
+    MultiSchemaResponse multiSchemaResponse = fetchMultiSchemaResponse(requestPath);
+    if (multiSchemaResponse == null) {
+      return Collections.emptyList();
+    }
+    List<SchemaEntry> valueSchemaEntryList = new ArrayList<>();
+    for (MultiSchemaResponse.Schema schema: multiSchemaResponse.getSchemas()) {
+      Schema writerSchema = preemptiveSchemaVerification(
+          AvroCompatibilityHelper.parse(schema.getSchemaStr()),
+          schema.getSchemaStr(),
+          schema.getId());
+      valueSchemaEntryList.add(new SchemaEntry(schema.getId(), writerSchema));
+    }
+    supersetSchemaIdAtomic.set(multiSchemaResponse.getSuperSetSchemaId());
+    return valueSchemaEntryList;
+  }
+
+  private boolean isValidSchemaEntry(SchemaEntry schemaEntry) {
+    return schemaEntry.getId() != SchemaData.INVALID_VALUE_SCHEMA_ID;
+  }
+
+  private boolean isValidSchemaId(int valueSchemaId) {
+    return valueSchemaId != SchemaData.INVALID_VALUE_SCHEMA_ID;
+  }
+
+  private MultiSchemaIdResponse fetchMultiSchemaIdResponse(String requestPath) throws VeniceClientException {
+    MultiSchemaIdResponse schemaIdResponse;
+    byte[] response = executeRouterRequest(requestPath);
+
+    if (response == null) {
```

Comment:
Can `response` be `null`? Or maybe it can be `null` if the response is `errored`, so shall we log the error message here?

Reply:
Got it.